src/diffusers/models/attention_processor.py

```diff
@@ -2336,28 +2390,85 @@ def __call__(
+            ):
+                ip_key = to_k_ip(current_ip_hidden_states)
+                ip_value = to_v_ip(current_ip_hidden_states)
+
+                ip_key = ip_key.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
+                ip_value = ip_value.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
+
+                # the output of sdp = (batch, num_heads, seq_len, head_dim)
+                # TODO: add support for attn.scale when we move to Torch 2.1
+                current_ip_hidden_states = F.scaled_dot_product_attention(
+                    query, ip_key, ip_value, attn_mask=None, dropout_p=0.0, is_causal=False
+                )
 
-            ip_key = ip_key.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
-            ip_value = ip_value.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
+                current_ip_hidden_states = current_ip_hidden_states.transpose(1, 2).reshape(
+                    batch_size, -1, attn.heads * head_dim
+                )
+                current_ip_hidden_states = current_ip_hidden_states.to(query.dtype)
+
+                seq_len = current_ip_hidden_states.shape[1]
+                o_h = masks[0].shape[1]
+                o_w = masks[0].shape[2]
+                ratio = o_w / o_h
+                mask_h = int(torch.sqrt(torch.tensor(seq_len / ratio)))
+                mask_h = int(mask_h) + int((seq_len % int(mask_h)) != 0)
+                mask_w = seq_len // mask_h
+
+                if len(mask.shape) == 2:
+                    mask = mask.unsqueeze(0)
+                mask_downsample = F.interpolate(
+                    torch.tensor(mask, dtype=torch.float32).unsqueeze(0), size=(mask_h, mask_w), mode="bicubic"
+                ).squeeze(0)
+
+                if mask_downsample.shape[0] < batch_size:
+                    mask_downsample = mask_downsample.repeat(batch_size, 1, 1)
+                if mask_downsample.shape[0] > batch_size:
+                    mask_downsample = mask_downsample[:batch_size, :, :]
+
+                mask_downsample = mask_downsample.view(mask_downsample.shape[0], -1, 1).repeat(
+                    1, 1, attn.heads * head_dim
+                )
 
-            # the output of sdp = (batch, num_heads, seq_len, head_dim)
-            # TODO: add support for attn.scale when we move to Torch 2.1
-            current_ip_hidden_states = F.scaled_dot_product_attention(
-                query, ip_key, ip_value, attn_mask=None, dropout_p=0.0, is_causal=False
-            )
+                mask_downsample = mask_downsample.to(query.dtype).to(current_ip_hidden_states.device)
 
-            current_ip_hidden_states = current_ip_hidden_states.transpose(1, 2).reshape(
-                batch_size, -1, attn.heads * head_dim
-            )
-            current_ip_hidden_states = current_ip_hidden_states.to(query.dtype)
+                current_ip_hidden_states = current_ip_hidden_states * mask_downsample
```

Comment:
I know, I didn't add checks on mask size yet. I think also ComfyUI implementation has the same issue, but I haven't tested it
 https://github.com/cubiq/ComfyUI_IPAdapter_plus/blob/90d3451cd970d5aa9cac55224e24a7c7fd98d253/IPAdapterPlus.py#L537

Reply:
I think it works with masks that aren't of the same ratio as the generation, is just not recommended. Maybe @cubiq can provide his insights here, I use the same code and it doesn't use the ratio, I think the `need checking` means that he wasn't completely sure of the formula used.

Reply:
Ok, I will check without ratio as in the other implementation! Thanks

Reply:
the attention mask is resized and stretched at each iteration, the aspect ratio doesn't matter but of course it's better if you provide the right size.

due to rounding error it might happen that you get the wrong size, but it's not very common and I think I have a solution for that already.

Reply:
I can confirm the issue is still there also with the other implementation

Reply:
In that case I really don't know what should be the best method of doing this that's consistent with diffusers.

In my case I prepare the mask latents outside the attention processor with the vae scale factor and the width and height of the generated image but it could be as simple as throwing an error telling the user that the masks must have the same aspect ratio than the generated image.